plugins/httpapi/nd.py

```diff
@@ -184,26 +194,63 @@ def send_request(self, method, path, data=None):
+            self.connection.set_option('persistent_command_timeout', self.params.get('timeout'))
+
         # Perform some very basic path input validation.
         path = str(path)
         if path[0] != '/':
             self.error = dict(code=self.status, message='Value of <path> does not appear to be formated properly')
             raise ConnectionError(json.dumps(self._verify_response(None, method, path, None)))
         full_path = self.connection.get_option('host') + path
         try:
-            self.connection.queue_message('vvvv', 'send_request() - connection.send({0}, {1}, {2}, {3})'.format(path, data, method, self.headers))
+            self.connection.queue_message('info', 'send_request() - connection.send({0}, {1}, {2}, {3})'.format(path, data, method, self.headers))
             response, rdata = self.connection.send(path, data, method=method, headers=self.headers)
-        except ConnectionError:
-            self.connection.queue_message('vvvv', 'login() - ConnectionError Exception')
+        except ConnectionError as connection_err:
+            self.connection.queue_message('info', 'send_request() - ConnectionError Exception: {0}'.format(connection_err))
             raise
         except Exception as e:
-            self.connection.queue_message('vvvv', 'send_request() - Generic Exception')
+            self.connection.queue_message('info', 'send_request() - Generic Exception: {0}'.format(e))
             if self.error is None:
                 self.error = dict(code=self.status, message='ND HTTPAPI send_request() Exception: {0} - {1}'.format(e, traceback.format_exc()))
             raise ConnectionError(json.dumps(self._verify_response(None, method, full_path, None)))
         return self._verify_response(response, method, full_path, rdata)
 
+    def send_file_request(self, method, path, file=None, data=None):
+        ''' This method handles all ND REST API requests other than login '''
+
+        self.error = None
+        self.path = ''
+        self.status = -1
+        self.info = {}
+        self.method = 'GET'
+        if method is not None:
+            self.method = method
+
+        try:
+            # create data field
+            data["uploadedFileName"] = os.path.basename(file)
+            data_str = io.StringIO()
+            json.dump(data, data_str)
+        except Exception as e:
+            self.error = dict(code=self.status, message='ND HTTPAPI create data field Exception: {0} - {1}'.format(e, traceback.format_exc()))
+            raise ConnectionError(json.dumps(self._verify_response(None, method, path, None)))
+
+        try: 
+            # create fields for MultipartEncoder
+            fields = dict(data=('data.json', data_str, 'application/json'), file=(os.path.basename(file), open(file, 'rb'), mimetypes.guess_type(file)))
+            mp_encoder = MultipartEncoder(fields=fields)
+            self.headers['Content-Type'] = mp_encoder.content_type
```

Comment:
Is this content_type needed for something special? It seems to cause issues for when a regular send_request() method is called after. The content-type is this multipart encoder type which results in a media type api error response. 

Managed to temporary solve this by hardcoding a reset of the content-type in the send_request() method but this is should probably be fixed differently.